### PR TITLE
Define policy for mulitchallenge enroll

### PR DESCRIPTION
**Pull request by cornelinux**
Each tokentype can define, if it can be enrolled
during /validate/check.
Prepare the policy to configure this.

Working on #2993

I marked this ready for review, since only the push implemenation is missing.
But if we start to review this,  maybe this can be merged soon, so that @pablo-knight  can build a dev release @nilsbehlen can use this for working on the implemenation on the client components.

### CODE REVIEW

File: doc/policies/authentication.rst [hunk 1]
```diff
@@ -361,6 +361,57 @@ In contrast to the generic AutoResync a user has to enter the token PIN only onc
...
+is displayed to the user. The user has to scan the QR code and will then have to enter the valid OTP value,
+generated by the newly scanned/enrolled token. Only after that, the user is finally authenticated.
+
+.. note:: 2step enrollment is currently not supported in this enrollment scenario.
+
+**SMS and Email**
+
+After the first successful authentication step the user is presented with an input field to
+enter his email address or mobile number. If done so, the user will then in the final step
+have to enter the OTP value sent via email or test message.
```
Comment (NNLemling) [inline doc/policies/authentication.rst:400]: *text message*

File: privacyidea/lib/config.py [hunk 2]
```diff
@@ -1012,3 +1012,13 @@ def import_config(data, name=None):
...
+
+
+def get_multichallenge_enrollable_tokentypes():
+    enrollable_tokentypes = []
+    # If the token is enrollable via multichallenge
+    for tclass in get_token_classes():
+        if tclass.is_multichallenge_enrollable:
+            enrollable_tokentypes.append(tclass.get_class_type())
+
+    return enrollable_tokentypes
```
Comment (NNLemling) [inline privacyidea/lib/config.py:1024]: Add newline at the end.

File: privacyidea/lib/tokenclass.py [hunk 3]
```diff
@@ -1582,16 +1589,22 @@ def check_challenge_response(self, user=None, passw=None, options=None):
...
+                        # Basically we have a successfully answered challenge
+                        otp_counter = 0
                     else:
-                        # increase the received_count
-                        challengeobject.set_otp_status()
+                        # Now see if the OTP matches:
+                        otp_counter = self.check_otp(passw, options=options)
+                        if otp_counter >= 0:
+                            # We found the matching challenge, so lets return the
+                            #  successful result and delete the challenge object.
```
Comment (NNLemling) [inline privacyidea/lib/tokenclass.py:1602]: Delete whitespace before **successful**

File: privacyidea/lib/tokens/smstoken.py [hunk 4]
```diff
@@ -328,20 +329,21 @@ def create_challenge(self, transactionid=None, options=None):
...
-                success, sent_message = self._send_sms(
-                    message=message_template, options=options)
-
-                # Create the challenge in the database
-                if is_true(get_from_config("sms.concurrent_challenges")):
-                    data = self.get_otp()[2]
-                else:
-                    data = None
+                data = None
+                # Only if this is NOT an multichallenge enrollment, we try to send the email
```
Comment (NNLemling) [inline privacyidea/lib/tokens/smstoken.py:337]: we try to send the >>**SMS**<<

**Review by NNLemling — COMMENTED** [review]
Please check my minor comments

File: privacyidea/lib/tokens/smstoken.py (hunk 4 above)
Comment (cornelinux) [reply privacyidea/lib/tokens/smstoken.py]: done in 45125f2f

File: privacyidea/lib/config.py (hunk 2 above)
Comment (cornelinux) [reply privacyidea/lib/config.py]: Done in 45125f2f

File: doc/policies/authentication.rst (hunk 1 above)
Comment (cornelinux) [reply doc/policies/authentication.rst]: Done in 45125f2f

File: privacyidea/lib/tokenclass.py [hunk 5]
```diff
@@ -1874,3 +1887,28 @@ def verify_enrollment(self, response):
...
+    def enroll_via_validate_2nd_step(self, passw, options=None):
+        """
+        This method is the optional second step of ENROLL_VIA_MULTICHALLENGE.
+        It is used in situations like the email token, sms token or push,
+        when enrollment via challenge response needs two steps.
+
+        :param options:
+        :return:
+        """
+        return True
```
Comment (NNLemling) [inline privacyidea/lib/tokenclass.py:1914]: Add newline at the end.

File: tests/test_api_push_validate.py [hunk 6]
```diff
@@ -355,3 +360,176 @@ def test_03_unfinished_enrolled_push_token(self):
...
+            self.assertEqual(result.get("authentication"), "CHALLENGE")
+            detail = res.json.get("detail")
+            transaction_id = detail.get("transaction_id")
+            self.assertTrue("Please scan the QR code!" in detail.get("message"), detail.get("message"))
+            # Get image and client_mode
+            self.assertEqual(CLIENTMODE.POLL, detail.get("client_mode"))
+            # Check, that multi_challenge is also contained.
+            self.assertEqual(CLIENTMODE.POLL, detail.get("multi_challenge")[0].get("client_mode"))
+            self.assertIn("image", detail)
+            #image = detail.get("image")
```
Comment (NNLemling) [inline tests/test_api_push_validate.py:454]: Remove unused commented out code.
`#image = detail.get("image")`

**Review by NNLemling — COMMENTED** [review]
Please check my minor comments.

File: tests/test_api_push_validate.py (hunk 6 above)
Comment (cornelinux) [reply tests/test_api_push_validate.py]: Done in 3f6ad403

**Review by NNLemling — APPROVED** [review]
Looks good!

**Pull request merged**